### PR TITLE
Add Arena::{with_capacity, capacity, reserve, clear}

### LINKED ISSUE #84: Add Arena::with_capacity
When creating large trees, inserting many nodes causes the underlying `Vec` to reallocate frequently.
`Arena::with_capacity` would simply create allocate the `Vec` beforehand.

```rust
fn with_capacity(capacity: usize) -> Self {
    Self {
        nodes: Vec::with_capacity(capacity),
        first_free_slot: None,
        last_free_slot: None,
    }
}
```
I would also suggest adding the `Arena::clear` function which simply clears the `Vec`, causing all the created `NodeId`s to be invalid.
This is equivalent to removing all the `Node`s while retaining the `Vec` capacity.
Unless I forgot something, I'm rather confident that such an implementation would not cause soundness issues, e.g. with the iterators.

### PR DESCRIPTION
**Pull request by slivering**
Closes #84

### CODE REVIEW

**Review by saschagrunert — CHANGES_REQUESTED** [review]
Thank you! Can you please add a bunch of test cases for the new API?

File: src/arena.rs [hunk 1]
```diff
@@ -49,6 +49,22 @@ impl<T> Arena<T> {
...
+    }
+
+    /// Reserves capacity for `additional` more nodes to be inserted.
+    ///
+    /// The arena may reserve more space to avoid frequent reallocations.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the new capacity exceeds isize::MAX bytes.
+    pub fn reserve(&mut self, additional: usize) {
```
Comment (slivering) [inline src/arena.rs:64]: For API coherence compared to `Vec`

**Comment by slivering** [comment]
I added a newline at the end of the files, should that fix the rustfmt check?

**Comment by slivering** [comment]
Ok, now I'm done

**Review by saschagrunert — APPROVED** [review]

**Pull request merged**